Issue 5. Add Items to list

sillytsundere opened the pull request:
## Description

These code changes enable users to add items to their shopping lists via a form on the ManageLists view. The form uses semantic html elements to adhere to accessibility standards. 

## Related Issue

Closes #5

## Acceptance Criteria

UI-related tasks:
- [x] The `ManageList` view displays a form that allows them to enter the name of the item and select how soon they anticipate needing to buy it again. There should be 3 choices for this:
	- “Soon”, corresponding to 7 days
	- “Kind of soon”, corresponding to 14 days
	- “Not soon”, corresponding to 30 days
- [x] The input that accepts the name of the item has a semantic `label` element associated with it
- [x] The user can submit this form with both the mouse and the `Enter` key  
- [x] When the user submits the form, they see a message indicating that the item either was or was not saved to the database.

Data-related tasks:
- [x] The `console.log` in the `addItem` function in `src/api/firebase.js`  is replaced with a function that adds the new document to the Firestore database. That function will be imported from the `firebase/firestore` module.
- [x]  The user’s soon/not soon/kind of soon choice is used to calculate `nextPurchasedDate`

## Type of Changes

enhancement

## Updates

### Before

![localhost_3000_manage-list (1)](https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/135680082/2bb385fc-062f-4f01-a19a-ff516e3096e2)

### After

![localhost_3000_manage-list](https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/135680082/bd421d04-1876-4a53-85b7-625a0ecc3e56)

## Testing Steps / QA Criteria

git pull
git checkout ja-pc-feature-add-item
npm ci
npm run start
Navigate to Manage List page to view form
Enter item and when you will next purchase item into form fields, submit form, view alert informing item was successfully saved
Navigate to List page to view newly added item

jaguilar89 committed:
edit ManageList.jsx

sillytsundere committed:
finished form on manage list view, refactored addItem function to save item to database on list

sillytsundere committed:
removed comment

rachelspencer committed:
enhance accessibility and fixform submission requirements

rachelspencer commented:
Hi @sillytsundere and @jaguilar89!
This looks so neat and tidy! I pulled this branch down and ran the app. 

I noticed I could submit an item when the input field and radio buttons were unselected. I could also submit an item with only a name and no date requirement, and also submit with no name and only a date requirement. When doing this I got the success alert.

To overcome this I added a `<required>` attribute to the `<input>` tags.

To enhance accessibility I also added an `<id>` tag who's value matches the `<forHTML>` attribute already existing on the `<label>` tags. 

I could enter an item using the mouse and enter key! I am not sure if you want to enhance the accessibility more and add functionality for a user to select the radio buttons by using the tab and enter key also. Would definitely be extra challenge but might be fun!

The `new item` const on line 19 isn't used, not sure if you want to take it out to clean up the `console.log`, or if you have plans to use it later.

Great work!!

PS - I'm not sure if we should be seeing the items in Firebase DB with the extra functionality added for this issue. I couldn't see them, but I'm also still getting the hang of Firebase.

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+						name="itemName"
+						type="text"
+						id="item"
+						required
```
added required so form field can't be left blank on submit

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+								name="daysUntilNextPurchase"
+								type="radio"
+								id="soon"
+								required
```
added required so form field can't be left blank on submit

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+								name="daysUntilNextPurchase"
+								type="radio"
+								id="kind-of-soon"
+								required
```
added required so form field can't be left blank on submit

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+								onChange={handleChangeDate}
+								name="daysUntilNextPurchase"
+								type="radio"
+								id="not-soon"
```
added `id` attribute to match `forHTML` attribute on `<label>` for accessibility

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+								onChange={handleChangeDate}
+								name="daysUntilNextPurchase"
+								type="radio"
+								id="kind-of-soon"
```
added `id` attribute to match `forHTML` attribute on `<label>` for accessibility

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+								onChange={handleChangeDate}
+								name="daysUntilNextPurchase"
+								type="radio"
+								id="soon"
```
added `id` attribute to match `forHTML` attribute on `<label>` for accessibility

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+						<li>
+							<label htmlFor="soon"> Soon (7 days)</label>
+							<input
+								value={7}
```
Consider making this your default value, to enhance accessibility. This can be achieved by adding another attribute here. If you chose to do this, you'll also need to update the initial state value to be "7".

jaguilar89 replied:
We initially leave this part of the form unselected to account for accidental form submissions where an item name would be provided but the first radio button would already be selected, before the user has a chance to select it on their own. Whereas if there is no default radio button value, the user would see an error message, asking them to select an option first. See below screenshot.

![image](https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/78065152/d966946d-7abc-4ccd-b272-dd63af4eaa3c)

As an alternative, we added a clear "Please select option" instruction to make it clear that the user must make a selection before submitting the form. See below screenshot.

![image](https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/78065152/6e9e8767-0fd3-4e61-8020-4d08b1cd13c9)

jaguilar89 replied:
@trushmi What do you think?

rachelspencer commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,90 @@
+	const handleSubmit = async (e) => {
+		e.preventDefault();
+		try {
+			const newItem = await addItem(listPath, {
```
The const `newItem` is being set here but not used. Maybe remove to clean up the console.log?

rachelspencer reviewed (approved):
Hey @jaguilar89 and @sillytsundere,

This looks so cool!

I added some comments on the changes I made on the branch, as well as a couple additional non-blocking suggestions that I didn't implement.

sillytsundere committed:
refactored code, simplified hooks into one handleSubmit hook, useState no longer necessary

sillytsundere commented:
Jose and I made more changes, we agree with the changes suggested and implemented by Rachel to make the items of the form required so as to prevent items with no name or next purchase date to be entered into the list. Also Jose refactored the code to eliminate the need for useState and refactored the multiple hooks into one handleSubmit hook. This also removes the const newItem, an unused variable, calling the await directly without saving it in a variable.

trushmi commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,75 @@
-		<p>
-			Hello from the <code>/manage-list</code> page!
-		</p>
+		<div>
```
this wrapper "div" seems unnecessary, can we remove it and just have <form ... starting here?

jaguilar89 replied:
Agreed. Div removed.

trushmi commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,75 @@
+		} catch (error) {
+			alert('There was a problem');
+		}
+		document.getElementById('item-form').reset();
```
Instead of selecting element by id, it's better to use useRef hook in React. Here is a link to documentation: https://react.dev/reference/react/useRef

I would recommend to write: 
```
import {useRef} from "react"; 
...
ManageList(...) {
   const formRef = useRef()
   
   const handleSubmit = async (e) => {
      ....
      formRef.current.reset()
   }
   ...
   
   return (
       <form ref={formRef} .... > 
       ...
       </form>
   )
}
```

jaguilar89 replied:
Good call! Change has been made.

trushmi commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,75 @@
-			Hello from the <code>/manage-list</code> page!
-		</p>
+		<div>
+			<form id="item-form" onSubmit={handleSubmit}>
```
I would recommend moving this form to a separate component in case we use it in the future. For example, we can create a component called RadioInput. The component should use props to receive the options, like this:

```
const options = [{label: "Soon (7 days)", value: 7, id: "soon"}, {label: "Kind of soon (14 days)", value: 14, id:"kind-of-soon"}];


...

<RadioInput name="daysUntilNextPurchase" options={options}/>
```
and inside of the component we can use those props:
```
function RadioInput({name, options}) {

  return (
    <ul>
      {options.map((option) => (
          <li>
             <label htmlFor={option.id}>{option.label}</label>
             <input value={option.value} name={name}... >
          </li>
        )
      )}
    <ul/>
  ) 
}
```

jaguilar89 replied:
This is a good suggestion. However, Paige and I think it's unnecessary for now simply because we don't think a radio button component will be needed in the future, based on the acceptance criteria of future issues. But, we can always keep this suggestion in mind in case it does become a necessary change in the future.

trushmi commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,75 @@
+import { addItem } from '../api/firebase';
+
+export function ManageList({ listPath }) {
+	async function handleSubmit(e) {
```
I would recommend to use here: 
```
const handleSubmit = async (e) => { ....
```
Instead of defining the function.

jaguilar89 replied:
Agreed, it's more readable this way. Change has been made.

trushmi reviewed (changes requested):
Looks good overall, but I think we can make a few improvements.

rachelspencer reviewed (approved):
@sillytsundere @jaguilar89 looks good!

jaguilar89 committed:
made suggested changes to ManageList.jsx

trushmi reviewed (approved):
Reviewed and tested. Looks good to me!

timothy-taylor commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,80 @@
+	const handleSubmit = async (e) => {
+		e.preventDefault();
+
+		const formData = new FormData(e.target);
```
Cool approach here using the FormData api in conjunction with default form/input functionality to access necessary data. You don't see this often in client-side React apps, but totally legit thing to do so that's cool to see.

timothy-taylor commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,80 @@
+			alert('There was a problem');
+		}
+
+		formRef.current.reset();
```
I wonder whether this `reset()` should be inside of the `try` branch of the `try ... catch ...`. If there is an error that means that the item wasn't saved, so we probably don't want to clear the form on people (in case they want to try again?)

jaguilar89 replied:
This makes sense, clearing a form after a successful submit is expected behavior, not clearing it would clearly indicate there was a problem. I added this change.

timothy-taylor commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,80 @@
+
+			alert('Item saved successfully');
+		} catch (error) {
+			alert('There was a problem');
```
We don't necessarily have to fully display the `error` in this catch statement, but I do think we could make this error status more specific.

jaguilar89 replied:
I modified the catch block to display a more detailed error message.

```javascript
try {
...
} catch (error) {
   alert(`There was a problem: ${error.message}`);
};

timothy-taylor commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,80 @@
+							name="daysUntilNextPurchase"
+							type="radio"
+							id="soon"
+							required
```
I said this to the other group as well, but the built in HTML validations actually aren't as accessible as one would hope. So I think down the road we should build our own validation + status displaying logic.

timothy-taylor commented on src/views/ManageList.jsx:
```diff
@@ -1,7 +1,80 @@
+import { useRef } from 'react';
+
+export function ManageList({ listPath }) {
+	const formRef = useRef();
```
the best practice recommendation by the React team is to declare this as `const formRef = useRef(null)` if you are going to assign the reference to a DOM element. I'm honestly not sure the different `null` vs `undefined` makes in the initial declaration in reality, but I would still recommend doing it so that your code can be as consistent as possible.

jaguilar89 replied:
Noted. I will add this change. Thanks!

timothy-taylor reviewed (approved):
Nice job completing this one! I left a few thoughts: feel to make any small changes you want or just make note of any changes you want to make moving forward (you can definitely leave the custom validation + status stuff for the future) and get this merged in!

jaguilar89 committed:
added suggested changes to ManageList.jsx; made null the initial value for useRef, moved form reset into try block so it only resets upon successful submission
